clients/python/test/e2e/ci/_data_classes.py

```diff
@@ -0,0 +1,172 @@
+    def check_consistency(self) -> "ClientConfig":
+        msg: str = (
+            f"Recieved osparc_client_version={self.osparc_client_version}, "
+            f"osparc_client_repo={self.osparc_client_repo}"
+            "and osparc_client_branch={self.osparc_client_branch}. "
+            "Either a version or a repo, branch pair must be specified. Not both."
+        )
+        # check at least one is empty
+        if not (
+            is_empty(self.osparc_client_version)
+            or (
+                is_empty(self.osparc_client_repo)
+                and is_empty(self.osparc_client_branch)
+            )
+        ):
+            raise ValueError(msg)
+        # check not both empty
+        if is_empty(self.osparc_client_version) and (
+            is_empty(self.osparc_client_repo) and is_empty(self.osparc_client_branch)
+        ):
+            raise ValueError(msg)
+        if is_empty(self.osparc_client_version):
+            if (
+                is_empty(self.osparc_client_repo)
+                or is_empty(self.osparc_client_branch)
+                or is_empty(self.osparc_client_workflow)
+                or is_empty(self.osparc_client_runid)
+            ):
+                raise ValueError(msg)
+        return self
+
+    @property
+    def version(self) -> Optional[str]:
+        return self.osparc_client_version
+
+    @property
+    def repo(self) -> Optional[str]:
+        return self.osparc_client_repo
+
+    @property
+    def branch(self) -> Optional[str]:
+        return self.osparc_client_branch
+
+    @property
+    def workflow(self) -> Optional[str]:
+        return self.osparc_client_workflow
+
+    @property
+    def runid(self) -> Optional[str]:
+        return self.osparc_client_runid
+
+    @property
+    def compatibility_ref(self) -> str:
+        """Returns the reference for this client in the compatibility table"""
+        if not is_empty(self.version):
+            return "production"
+        else:
+            assert isinstance(self.branch, str)
+            return self.branch
+
+    @property
+    def client_ref(self) -> str:
+        """Returns a short hand reference for this client"""
+        if not is_empty(self.version):
+            assert isinstance(self.version, str)
+            return self.version
+        else:
+            assert isinstance(self.branch, str)
+            return self.branch
+
+
+class PytestConfig(BaseModel):
+    """Holds the pytest configuration
+    N.B. paths are relative to clients/python/test/e2e
+    """
+
+    env: str
+    required_plugins: str
+    addopts: str
+
+
+class Artifacts(BaseModel):
+    artifact_dir: Path
+    result_data_frame: Path
+    log_dir: Path
+
+
+class PytestIniFile(BaseModel):
+    """Model for validating the .ini file"""
+
+    pytest: PytestConfig
+    client: ClientConfig
+    server: ServerConfig
+    artifacts: Artifacts
+
+    @classmethod
+    def read(cls, pth: Path = _PYTEST_INI) -> "PytestIniFile":
```

Comment:
Hmm, interesting. Yes, perhaps I could have used these pydantic settings.
The reasoning I had is that I am not such a big fan of environment variables because I feel it makes it harder to reproduce everything. But perhaps that feeling is not justified :man_shrugging:. My design decision was that everything needed to run the test has to go into the `pytest.ini`. Also environment variables. That makes it easy to reproduce because I just need to get that file from the artifacts and install the needed python packages. The way environment variables are exposed is handled automatically by the `pytest-env` plugin.